app/src/swap/handle_sign_transaction.c

```diff
@@ -0,0 +1,187 @@
+    // input {0xEE, 0x00, 0xFF} should be stored like {0x00, 0x00, 0x00, 0x00, 0x00, 0xEE, 0x00, 0xFF}
+    memcpy(amount + 16 - sign_transaction_params->amount_length, sign_transaction_params->amount,
+           sign_transaction_params->amount_length);
+
+    memcpy(fees + 8 - sign_transaction_params->fee_amount_length, sign_transaction_params->fee_amount,
+           sign_transaction_params->fee_amount_length);
+
+    amount_length = sign_transaction_params->amount_length;
+
+    // Full reset the global variables
+    os_explicit_zero_BSS_segment();
+
+    // Keep the address at which we'll reply the signing status
+    G_swap_sign_return_value_address = &sign_transaction_params->result;
+
+    // Commit the values read from exchange to the clean global space
+    G_swap_state.amount_length = amount_length;
+    memcpy(G_swap_state.amount, amount, sizeof(amount));
+    memcpy(G_swap_state.destination_address, destination_address, sizeof(G_swap_state.destination_address));
+    readU64BE(fees, &G_swap_state.fees);
+
+    return true;
+}
+
+// Ensure the received transaction matches what was validated in the Exchange app UI
+parser_error_t check_swap_conditions(parser_tx_t *txObj) {
+    parser_error_t err = parser_unexpected_error;
+    if (txObj == NULL) {
+        return err;
+    }
+    // Check transaction method arguments number. Balance transfer Should be 3 (for tx v26).
+    // [dest(address type) |  dest(address) | value(amount)]
+    // We will check that index 5 does not have the TIP
+    if (txObj->blob.totalMethodItems != SWAP_EXPECTED_ITEMS && txObj->blob.totalMethodItems != SWAP_EXPECTED_ITEMS + 1) {
+        ZEMU_LOGF(100, "Wrong swap tx method arguments count %d.\n", txObj->blob.totalMethodItems);
+        return parser_swap_tx_wrong_method_args_num;
+    }
+    // Check network.
+    const char *valid_network = "polkadot";
+    char tmpKey[20] = {0};
+    char tmpValue[65] = {0};
+
+    uint8_t pageCount = 0;
+    ui_field_t uiFields = {.displayIdx = 0,
+                           .outKey = tmpKey,
+                           .outKeyLen = sizeof(tmpKey),
+                           .outVal = tmpValue,
+                           .outValLen = sizeof(tmpValue),
+                           .pageIdx = 0,
+                           .pageCount = &pageCount};
+
+    CHECK_ERROR(parser_getItem(txObj, &uiFields));
+    if (strncmp(valid_network, tmpValue, strlen(valid_network) + 1) != 0) {
+        ZEMU_LOGF(200, "Swap not enable on %s network.\n", tmpValue);
+        return parser_swap_tx_wrong_method;
+    }
+
+    // Check method.
+    uiFields.displayIdx = 1;
+    MEMZERO(tmpKey, sizeof(tmpKey));
+    MEMZERO(tmpValue, sizeof(tmpValue));
+    const char *valid_tx_pallet = "Balances";
+    const char *valid_tx_call = "transfer_allow_death";
+
+    CHECK_ERROR(parser_getItem(txObj, &uiFields));
+    if (strncmp(valid_tx_pallet, tmpKey, strlen(valid_tx_pallet) + 1) != 0 ||
```

Comment:
Here you're checking `tmpKey` value. Why checking `tmpKey` value is not needed in the other cases?

Reply:
Not checked for address and destination because it's only the format of amount or destination
If the format is wrong the check will fail anyway (that's how I understand it, it's code from Zondax)

Reply:
Oh yes, looks good to me then!